letsencrypt/plugins/webroot.py

```diff
@@ -60,23 +60,38 @@ def prepare(self):  # pylint: disable=missing-docstring
             logger.debug("Creating root challenges validation dir at %s",
                          self.full_roots[name])
             try:
-                os.makedirs(self.full_roots[name])
-                # Set permissions as parent directory (GH #1389)
-                # We don't use the parameters in makedirs because it
-                # may not always work
+                # Change the permissiosn to be writable (GH #1389)
+                # We also set umask because os.makedirs's mode parameter does
+                # not always work:
                 # https://stackoverflow.com/questions/5231901/permission-problems-when-creating-a-dir-with-os-makedirs-python
+                # We set the umask instead of going the chmod route to ensure the client
+                # can also run as non-root (GH #1795)
+
                 stat_path = os.stat(path)
-                filemode = stat.S_IMODE(stat_path.st_mode)
-                os.chmod(self.full_roots[name], filemode)
-                # Set owner and group, too
-                os.chown(self.full_roots[name], stat_path.st_uid,
-                          stat_path.st_gid)
+                old_umask = os.umask(0o022)
```

Comment:
`old_umask` should be obtained outside the try block.  If anything up to and including this line throws an exception, `os.umask(old_umask)` in the `finally` block is otherwise called with an undefined value.